overlord/hookstate/ctlcmd/ctlcmd.go

```diff
@@ -113,47 +113,50 @@ func (f ForbiddenCommandError) Error() string {
 	return f.Message
 }
 
-// ForbiddenCommand contains information about an attempt to use a command in a context where it is not allowed.
-type ForbiddenCommand struct {
-	Uid  uint32
-	Name string
-}
-
-func (f *ForbiddenCommand) Execute(args []string) error {
-	return &ForbiddenCommandError{Message: fmt.Sprintf("cannot use %q with uid %d, try with sudo", f.Name, f.Uid)}
-}
-
 // nonRootAllowed lists the commands that can be performed even when snapctl
 // is invoked not by root.
 var nonRootAllowed = []string{"get", "services", "set-health", "is-connected", "system-mode"}
 
 // Run runs the requested command.
 func Run(context *hookstate.Context, args []string, uid uint32) (stdout, stderr []byte, err error) {
+	if len(args) == 0 {
+		return nil, nil, fmt.Errorf("snapctl cannot run without args")
+	}
+
+	if !isAllowedToRun(args, uid, args[0]) {
+		return nil, nil, &ForbiddenCommandError{Message: fmt.Sprintf("cannot use %q with uid %d, try with sudo", args[0], uid)}
```

Comment:
I don't think a hook can actually run with sudo or a snap app can call it internally. It's probably best to drop the sudo bit.
```suggestion
		return nil, nil, &ForbiddenCommandError{Message: fmt.Sprintf("cannot use %q with uid %d", args[0], uid)}
```

Reply:
IMHO this message fine as is - hooks are run by snapd so it's always uid 0; if it's called by an app then this hints the user about the app needing sudo as long as the error surfaces from the app?

Reply:
What I'm aiming at is that IIRC none of the apparmor rules allow running sudo by a strictly confined app.

Reply:
Yes that's correct but what I mean is hinting the user that the app needs uid 0, so `$ sudo theapp` (done by the user) would make it succeed with running snapctl internally.